### PR TITLE
Initial tables

### PR DESCRIPTION
**Pull request by luisffc**
Initial setup + added the following tables:

- auth0_client
- auth0_connection
- auth0_hook
- auth0_log
- auth0_organization
- auth0_role
- auth0_signing_key
- auth0_stats_daily
- auth0_tenant_settings
- auth0_user

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,79 @@
...
+
+```sql
+select
+  email,
+  id,
+  updated_at
+from
+  auth0_user
+where
+  not email_verified
```
Comment (misraved) [inline README.md]:
```suggestion
  not email_verified;
```

File: auth0/table_auth0_action.go [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+
+	client, err := Connect(ctx, d)
+	if err != nil {
+		logger.Error("auth0_action.getAuth0Actions", "connect_error", err)
+		return nil, err
+	}
+
+	actionsResponse, err := client.Action.Read(id)
+	if err != nil {
+		logger.Error("auth0_action.getAuth0Actions", "get_actions_error", err)
```
Comment (misraved) [inline auth0/table_auth0_action.go]:
```suggestion
		logger.Error("auth0_action.getAuth0Actions", "query_error", err)
```

File: auth0/table_auth0_client.go [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+	"github.com/turbot/steampipe-plugin-sdk/v5/plugin"
+	"github.com/turbot/steampipe-plugin-sdk/v5/plugin/transform"
+)
+
+//// TABLE DEFINITION
+
+func tableAuth0Client() *plugin.Table {
+	return &plugin.Table{
+		Name:        "auth0_client",
+		Description: "Client is an application or a sso integration.",
```
Comment (misraved) [inline auth0/table_auth0_client.go]:
```suggestion
		Description: "Client is an application or an SSO integration.",
```

File: docs/tables/auth0_action.md [hunk 4]
```diff
@@ -0,0 +1,60 @@
+# Table: auth0_action
+
+Actions are secure, tenant-specific, versioned functions written in Node.js that execute at certain points within the Auth0 platform.
+
+## Examples
+
+
```
Comment (misraved) [inline docs/tables/auth0_action.md]:
```suggestion
```

File: docs/tables/auth0_action.md [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+select
+  id,
+  name,
+  supported_triggers,
+  updated_at
+from
+  auth0_action
+where
+  runtime = 'node12' and
+  all_changes_deployed
```
Comment (misraved) [inline docs/tables/auth0_action.md]:
```suggestion
  all_changes_deployed;
```

File: docs/tables/auth0_action.md [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+
+```sql
+select
+  id,
+  name,
+  updated_at
+from
+  auth0_action
+where
+  supported_triggers -> 0 ->> 'id' = 'post-change-password'
```
Comment (misraved) [inline docs/tables/auth0_action.md]:
```suggestion
  supported_triggers -> 0 ->> 'id' = 'post-change-password';
```

File: docs/tables/auth0_action.md [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+
+### Action code by name
+
+```sql
+select
+  code
+from
+  auth0_action
+where
+  name = 'send-notification'
```
Comment (misraved) [inline docs/tables/auth0_action.md]:
```suggestion
  name = 'send-notification';
```

File: docs/tables/auth0_action.md [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+select
+  id,
+  name,
+  deployed_version ->> 'number' version,
+  supported_triggers,
+  updated_at
+from
+  auth0_action
+where
+  all_changes_deployed
```
Comment (misraved) [inline docs/tables/auth0_action.md]:
```suggestion
  all_changes_deployed;
```

File: docs/tables/auth0_client.md [hunk 9]
```diff
@@ -0,0 +1,44 @@
+# Table: auth0_client
+
+Client is an application or a sso integration.
+
+## Examples
+
+
```
Comment (misraved) [inline docs/tables/auth0_client.md]:
```suggestion

```

File: docs/tables/auth0_connection.md [hunk 10]
```diff
@@ -0,0 +1,87 @@
+# Table: auth0_connection
+
+Connection is the relationship between Auth0 and a source of users.
+
+## Examples
+
+
```
Comment (misraved) [inline docs/tables/auth0_connection.md]:
```suggestion

```

File: docs/tables/auth0_connection.md [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+  id,
+  name,
+  strategy
+from
+  auth0_connection
+where
+  options -> 'mfa' ->> 'active' = 'true';
+```
+
+### Connections with sign up disabled
```
Comment (misraved) [inline docs/tables/auth0_connection.md]:
```suggestion
### Connections with sign-up disabled
```

File: docs/tables/auth0_hook.md [hunk 12]
```diff
@@ -0,0 +1,51 @@
+# Table: auth0_hook
+
+Hooks allow you to customize the behavior of Auth0 with Node.js code that is executed in selected extension points.
+
+## Examples
+
+
```
Comment (misraved) [inline docs/tables/auth0_hook.md]:
```suggestion

```

File: docs/tables/auth0_log.md [hunk 13]
```diff
@@ -0,0 +1,69 @@
+# Table: auth0_log
+
+Logs are all the events that occur in your tenants including user authentication and administrative actions such as adding and updating applications, connections, and rules.
+Log type codes: https://auth0.com/docs/deploy-monitor/logs/log-event-type-codes
+
+## Examples
+
+
```
Comment (misraved) [inline docs/tables/auth0_log.md]:
```suggestion

```

File: docs/tables/auth0_organization.md [hunk 14]
```diff
@@ -0,0 +1,36 @@
+# Table: auth0_organization
+
+Organization is used to allow B2B customers to better manage their partners and customers, and to customize the ways that end-users access their applications.
+
+## Examples
+
+
```
Comment (misraved) [inline docs/tables/auth0_organization.md]:
```suggestion

```

File: docs/tables/auth0_role.md [hunk 15]
```diff
@@ -0,0 +1,43 @@
+# Table: auth0_role
+
+Role is a collection of permissions that you can apply to user.
+
+## Examples
+
```
Comment (misraved) [inline docs/tables/auth0_role.md:6]:
```suggestion
```

File: docs/tables/auth0_role_assigned_user.md [hunk 16]
```diff
@@ -0,0 +1,47 @@
+# Table: auth0_role_assigned_user
+
+List of users assigned to role.
+
+## Examples
+
+
```
Comment (misraved) [inline docs/tables/auth0_role_assigned_user.md]:
```suggestion

```
Please update all the table docs having the above pattern of occurrence.

File: docs/tables/auth0_signing_key.md [hunk 17]
```diff
@@ -0,0 +1,73 @@
...
+  thumbprint,
+  current_since,
+  current_until
+from
+  auth0_signing_key
+where
+  previous;
+```
+
+### Average time previous signing key had been available
```
Comment (misraved) [inline docs/tables/auth0_signing_key.md]:
```suggestion
### Average time for which the previous signing keys were available
```

File: docs/tables/auth0_stats_daily.md [hunk 18]
```diff
@@ -0,0 +1,44 @@
+# Table: auth0_stats_daily
+
+Daily stats for an Auth0 Tenant.
+
+## Examples
+
+
+### Top 5 days with highest number of sign ups
```
Comment (misraved) [inline docs/tables/auth0_stats_daily.md]:
```suggestion
### Top 5 days with the highest number of sign-ups
```

File: docs/tables/index.md [hunk 19]
```diff
@@ -0,0 +1,106 @@
...
+
+```sql
+select
+  email,
+  id,
+  updated_at
+from
+  auth0_user
+where
+  not email_verified
```
Comment (misraved) [inline docs/tables/index.md]:
```suggestion
  not email_verified;
```
Please add the missing `;` at the end of query examples.

File: config/auth0.spc [hunk 20]
```diff
@@ -0,0 +1,12 @@
+connection "auth0" {
```
Comment (misraved) [inline config/auth0.spc:1]: @luisffc could you please mention which arguments in the connection config file are `Required` fields and which ones are `Optional`?

File: auth0/table_auth0_action.go [hunk 21]
```diff
@@ -0,0 +1,104 @@
...
+			KeyColumns: plugin.SingleColumn("id"),
+		},
+
+		Columns: []*plugin.Column{
+			{Name: "id", Description: "A unique ID for the action.", Type: proto.ColumnType_STRING, Transform: transform.FromField("ID")},
+			{Name: "name", Description: "The name of the action.", Type: proto.ColumnType_STRING},
+			{Name: "supported_triggers", Description: "List of triggers that this action supports. At this time, an action can only target a single trigger at a time.", Type: proto.ColumnType_JSON},
+			{Name: "code", Description: "The source code of the action.", Type: proto.ColumnType_STRING},
+			{Name: "dependencies", Description: "List of third party npm modules, and their versions, that this action depends on.", Type: proto.ColumnType_JSON},
+			{Name: "runtime", Description: "The Node runtime. For example `node16`, defaults to `node12`", Type: proto.ColumnType_STRING},
```
Comment (misraved) [inline auth0/table_auth0_action.go]:
```suggestion
			{Name: "runtime", Description: "The Node runtime. For example `node16`, defaults to `node12`.", Type: proto.ColumnType_STRING},
```

File: auth0/table_auth0_action.go [hunk 22]
```diff
@@ -0,0 +1,104 @@
...
+		return nil, err
+	}
+
+	var pageNumber int
+	for {
+		actionsResponse, err := client.Action.List(
+			management.Page(pageNumber),
+		)
+		if err != nil {
+			logger.Error("auth0_action.listAuth0Actions", "list_actions_error", err)
```
Comment (misraved) [inline auth0/table_auth0_action.go]:
```suggestion
			logger.Error("auth0_action.listAuth0Actions", "query_error", err)
```
@luisffc could we make this update across all the tables to maintain uniformity across all the plugins?

File: auth0/table_auth0_action.go [hunk 23]
```diff
@@ -0,0 +1,104 @@
...
+			return nil, err
+		}
```
Comment (misraved) [inline auth0/table_auth0_action.go:62]: The table code seems to be missing the limit parameter check. While going through the documentation, I came across the `per_page` parameter, could we use this parameter to check against the query limit or set max items to be returned per page?

https://auth0.com/docs/api/management/v2#!/Actions/get_actions

File: auth0/table_auth0_client.go [hunk 24]
```diff
@@ -0,0 +1,128 @@
...
+		}
+		pageNumber = pageNumber + 1
+	}
+
+	return nil, err
+}
+
+//// GET FUNCTION
+
+func getAuth0Clients(ctx context.Context, d *plugin.QueryData, _ *plugin.HydrateData) (interface{}, error) {
```
Comment (misraved) [inline auth0/table_auth0_client.go]:
```suggestion
func getAuth0Client(ctx context.Context, d *plugin.QueryData, _ *plugin.HydrateData) (interface{}, error) {
```

**Review by misraved — COMMENTED** [review]
@luisffc the initial structure of the table looks great. I have left feedback comments which are mainly around the following points:
- Can we define the max limit parameter for each table to define the `limit` field? Please check if `per_page` parameter can be used to define the limit.
- The `Get Function` should be defined in `singular` form to maintain consistency across all the plugins. For instance `getAuth0Client` instead of `getAuth0Clients`.
- While logging, I think it would be beneficial to use `query_error` to log the API failures to maintain consistency across all the plugins.

File: auth0/table_auth0_connection.go [hunk 25]
```diff
@@ -0,0 +1,102 @@
...
+			Hydrate:    getAuth0Connections,
+			KeyColumns: plugin.SingleColumn("id"),
+		},
+
+		Columns: []*plugin.Column{
+			{Name: "id", Description: "A generated string identifying the connection.", Type: proto.ColumnType_STRING, Transform: transform.FromField("ID")},
+			{Name: "name", Description: "The name of the connection.", Type: proto.ColumnType_STRING},
+			{Name: "display_name", Description: "The display name of the connection.", Type: proto.ColumnType_STRING},
+			{Name: "strategy", Description: "The identity provider identifier for the connection.", Type: proto.ColumnType_STRING},
+			{Name: "is_domain_connection", Description: "True if the connection is domain level", Type: proto.ColumnType_BOOL},
```
Comment (misraved) [inline auth0/table_auth0_connection.go]:
```suggestion
			{Name: "is_domain_connection", Description: "True if the connection is domain level.", Type: proto.ColumnType_BOOL},
```

File: auth0/table_auth0_log.go [hunk 26]
```diff
@@ -0,0 +1,112 @@
...
+			{Name: "user_id", Description: "ID of the user involved in the log event.", Type: proto.ColumnType_STRING, Transform: transform.FromField("UserID")},
+			{Name: "user_name", Description: "Name of the user involved in the log event.", Type: proto.ColumnType_STRING},
+			{Name: "user_agent", Description: "User agent string from the client device that caused the event.", Type: proto.ColumnType_STRING},
+			{Name: "audience", Description: "API audience the event applies to.", Type: proto.ColumnType_STRING},
+			{Name: "scope", Description: "Scope permissions applied to the event.", Type: proto.ColumnType_STRING},
+			{Name: "strategy", Description: "Name of the strategy involved in the event.", Type: proto.ColumnType_STRING},
+			{Name: "strategy_type", Description: "Type of strategy involved in the event.", Type: proto.ColumnType_STRING},
+			{Name: "is_mobile", Description: "Whether the client was a mobile device (true) or desktop/laptop/server (false).", Type: proto.ColumnType_BOOL},
+			{Name: "location_info", Description: "Information about the location that triggered this event based on the `ip`.", Type: proto.ColumnType_JSON},
```
Comment (misraved) [inline auth0/table_auth0_log.go]:
```suggestion
			{Name: "location_info", Description: "Information about the location that triggered this event based on the `IP`.", Type: proto.ColumnType_JSON},
```

File: auth0/table_auth0_organization.go [hunk 27]
```diff
@@ -0,0 +1,95 @@
...
+		List: &plugin.ListConfig{
+			Hydrate: listAuth0Organizations,
+		},
+		Get: &plugin.GetConfig{
+			Hydrate:    getAuth0Organizations,
+			KeyColumns: plugin.SingleColumn("id"),
+		},
+
+		Columns: []*plugin.Column{
+			{Name: "id", Description: "Organization identifier", Type: proto.ColumnType_STRING, Transform: transform.FromField("ID")},
```
Comment (misraved) [inline auth0/table_auth0_organization.go]:
```suggestion
			{Name: "id", Description: "Organization identifier.", Type: proto.ColumnType_STRING, Transform: transform.FromField("ID")},
```

**Review by misraved — COMMENTED** [review]
@luisffc Please take a look at the additional review comments 👍.
A couple of questions that might need attention:
- Does the API return an empty row if the required qual is invalid or does not exist? If that is not the case, should we define an `IgnoreConfig` to return an empty row?
- Does the API return any rate limit errors? I see that the API does mention the `429` error code as a part of the response - https://auth0.com/docs/api/management/v2#!/Actions/get_actions. Should we add a separate section to handle it, like we have handled it in https://github.com/turbot/steampipe-plugin-jira/pull/72

File: auth0/table_auth0_action.go (hunk 23 above)
Comment (luisffc) [reply auth0/table_auth0_action.go:62]: yep, this is the "limit" parameter

File: config/auth0.spc (hunk 20 above)
Comment (luisffc) [reply config/auth0.spc:1]: changed, thanks

**Comment by luisffc** [comment]
Thanks for the review @misraved 
I've applied and pushed the fixes and improvements

**Pull request merged**